src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPECompilerManager.cs

```diff
@@ -0,0 +1,372 @@
+                    foreach (string file in _designTimeInputs)
+                    {
+                        ImmutableInterlocked.TryAdd(ref _filesToCompile, file, /* ignoreWriteTime */ false);
+                    }
+                    // Since we've just queued every file, we don't care about any other changes
+                    break;
+                }
+                else
+                {
+                    // A normal design time input, so just add it to the queue
+                    ImmutableInterlocked.TryAdd(ref _filesToCompile, changedFile, /* ignoreWriteTime */ false);
+                }
+            }
+
+            QueueCompilation();
+        }
+
+        internal void ProcessDataflowChanges(IProjectVersionedValue<Tuple<DesignTimeInputs, IProjectSubscriptionUpdate>> input)
+        {
+            DesignTimeInputs inputs = input.Value.Item1;
+
+            IProjectChangeDescription configChanges = input.Value.Item2.ProjectChanges[ConfigurationGeneral.SchemaName];
+
+            // On the first call where we receive design time inputs we queue compilation of all of them, knowing that we'll only compile if the file write date requires it
+            if (_designTimeInputs == null)
+            {
+                AddAllInputsToQueue(false);
+            }
+            else
+            {
+                // If its not the first call...
+
+                // If a new shared design time input is added, we need to recompile everything regardless of source file modified date
+                // because it could be an old file that is being promoted to a shared input
+                if (inputs.SharedInputs.Except(_sharedDesignTimeInputs, StringComparers.Paths).Any())
+                {
+                    AddAllInputsToQueue(true);
+                }
+                // If the namespace or output path inputs have changed, then we recompile every file regardless of date
+                else if (configChanges.Difference.ChangedProperties.Contains(ConfigurationGeneral.RootNamespaceProperty) ||
+                         configChanges.Difference.ChangedProperties.Contains(ConfigurationGeneral.ProjectDirProperty) ||
+                         configChanges.Difference.ChangedProperties.Contains(ConfigurationGeneral.IntermediateOutputPathProperty))
+                {
+                    AddAllInputsToQueue(true);
+                }
+                else
+                {
+                    // Otherwise we just queue any new design time inputs, and still do date checks
+                    foreach (string file in inputs.Inputs.Except(_designTimeInputs, StringComparers.Paths))
+                    {
+                        ImmutableInterlocked.TryAdd(ref _filesToCompile, file, /* ignoreWriteTime */ false);
+                    }
+                }
+            }
+
+            // Make sure we have the up to date list of inputs
+            _designTimeInputs = inputs.Inputs;
+            _sharedDesignTimeInputs = inputs.SharedInputs;
```

Comment:
These fields are not updated atomically which is a problem because ProcessFileChangeNotification and ProcessDataflowChanges can overlap, causing the former to see incomplete state.